### PR TITLE
New tests for matrixFileHandler and hiCMatrix

### PR DESCRIPTION
**Pull request by Rungetf**
Hi everyone,

I wrote some little tests for the matrixFileHandler and started testing on hiCMatrix only accepting `.cool` and `.h5` formats.

I'll extend the tests soon, but just to check that loading or saving with the FileHandler is working I thought I open a pull request...

@joachimwolff Am I correct that the __init__() of hiCMatrix doesn't check for any other format than `.cool` and `.h5`? So currently it accepts all formats but fails with others than `.cool` and `.h5`. Anyway I marked the test with XFail, because other formats cause an error.

@joachimwolff Can you please have a short look at the tests just to see if that's more or less how you expected the tests to be. I really only tested the runability of the load (and save if possible). But I'll come to your office tomorrow anyways...

### CODE REVIEW

File: hicmatrix/test/test_HiCMatrix.py [hunk 1]
```diff
@@ -1499,3 +1584,86 @@ def test_intervalListToIntervalTree(capsys):
...
+
+    hic.matrix = csr_matrix(matrix)
+    obs_exp_matrix = hic.convert_to_obs_exp_matrix(maxdepth=20).todense()
+    test_matrix = np.array([[1., 0.8, 1., 0., 0.],
+                            [0., 4., 1.5, 1., 0.],
+                            [0., 0., 0., 0.7, np.nan],
+                            [0., 0., 0., 0., np.nan],
+                            [0., 0., 0., 0., 0.]])
+
+    nt.assert_equal(obs_exp_matrix, test_matrix)
```
Comment (joachimwolff) [inline hicmatrix/test/test_HiCMatrix.py]: Maybe use for floating values `assert_almost_equal` with a delta of 2. 
https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_almost_equal.html

File: hicmatrix/test/test_matrixFileHandler.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import pytest
+
+ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), "test_data/")
+outfile = '/tmp/matrix'
+
+# TODO(Frederic): add more testcases eg., more parameters for file handler, different parameters for saving ...
+
+
+def test_load_and_save_homer(capsys):
+    homer_outfile = outfile + '.homer'
```
Comment (joachimwolff) [inline hicmatrix/test/test_matrixFileHandler.py]: Please split it in `test_load` and `test_save` and mark `test_save` as xfail.

File: hicmatrix/test/test_matrixFileHandler.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    fh.set_matrix_variables(matrix, cut_intervals, nan_bins, correction_factors, distance_counts)
+    # and save it.
+    with pytest.raises(TypeError):
+        fh.save(pName=homer_outfile, pSymmetric=False, pApplyCorrection=False)  # not implemented
+        os.unlink(homer_outfile)
+
+
+def test_load_and_save_h5():
+    h5_outfile = outfile + '.h5'
+
```
Comment (joachimwolff) [inline hicmatrix/test/test_matrixFileHandler.py]: Same here.

File: hicmatrix/test/test_matrixFileHandler.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    fh.set_matrix_variables(matrix, cut_intervals, nan_bins, correction_factors, distance_counts)
+    # and save it.
+    fh.save(h5_outfile, True, None)
+
+    os.unlink(h5_outfile)
+
+
+def test_load_and_save_hicpro():
+    hicpro_outfile = outfile + '.hicpro'
+
```
Comment (joachimwolff) [inline hicmatrix/test/test_matrixFileHandler.py]: And here.

File: hicmatrix/test/test_matrixFileHandler.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    # set matrix variables
+    fh.set_matrix_variables(matrix, cut_intervals, nan_bins, correction_factors, distance_counts)
+    # and save it.
+    fh.save(pName=cool_outfile, pSymmetric=True, pApplyCorrection=True)
+
+    os.unlink(cool_outfile)
+
+
+def test_load_and_save_ginteractions():
+    pass
```
Comment (joachimwolff) [inline hicmatrix/test/test_matrixFileHandler.py]: As above but mark the load function as xfail.

File: hicmatrix/test/test_HiCMatrix.py [hunk 6]
```diff
@@ -1499,3 +1584,86 @@ def test_intervalListToIntervalTree(capsys):
...
+    test_matrix = np.array([[0, 10, 20, 30, -1],
+                            [0, 0, 10, 20, -1],
+                            [0, 0, 0, 10, -1],
+                            [0, 0, 0, 0, -1],
+                            [0, 0, 0, 0, 0]], dtype='i4')
+    nt.assert_equal(matrix, test_matrix)
+
+    chrom_list = chrom_list.tolist()
+    test_chrom_list = ['a', 'a', 'a', 'a', '', 'a', 'a', 'a', '', 'a', 'a', '', 'a',
+                       '', 'b']
```
Comment (joachimwolff) [inline hicmatrix/test/test_HiCMatrix.py]: Please ignore the line length linting. You can exclude this rule in your linter.

**Comment by joachimwolff** [comment]
> @joachimwolff Am I correct that the init() of hiCMatrix doesn't check for any other format than .cool and .h5? So currently it accepts all formats but fails with others than .cool and .h5. Anyway I marked the test with XFail, because other formats cause an error.

Correct it only accepts cool and h5, for all others it should fail.

**Comment by Rungetf** [comment]
@joachimwolff I split the tests into seperate tests for saving and loading and marked some of them with Xfail. I don't think there is a testfile for the ginteraction format so I skipped this test at the moment. I'll implement it as soon as possible.

@joachimwolff I also changed the equallity test for the float matrices in the hicmatrix test to an almost_equal test as you suggested.

**Comment by joachimwolff** [comment]
@Rungetf  One test case passed but was marked as xfail. Please check what is going on and update the test case.

File: hicmatrix/test/test_matrixFileHandler.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+def test_load_h5(capsys):
+    # create matrixFileHandler instance with filetype 'h5'
+    pMatrixFile = ROOT + 'Li_et_al_2015.h5'
+    fh = MatrixFileHandler(pFileType='h5', pMatrixFile=pMatrixFile)
+    assert fh is not None
+
+    # load data
+    matrix, cut_intervals, nan_bins, distance_counts, correction_factors = fh.load()
+
+    # TODO(Frederic): figure out how to test matrix. first printed values are below.
```
Comment (joachimwolff) [inline hicmatrix/test/test_matrixFileHandler.py]: Solved?

File: hicmatrix/test/test_matrixFileHandler.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    pMatrixFile = ROOT + 'test_matrix.hicpro'
+    pBedFileHicPro = ROOT + 'test_matrix.bed'
+    fh = MatrixFileHandler(pFileType='hicpro', pMatrixFile=pMatrixFile, pBedFileHicPro=pBedFileHicPro)
+    assert fh is not None
+
+    # load data
+    matrix, cut_intervals, nan_bins, distance_counts, correction_factors = fh.load()
+
+    # TODO(Frederic): figure out how to test matrix.
+    # (0, 0)	41.345793
```
Comment (joachimwolff) [inline hicmatrix/test/test_matrixFileHandler.py]: Solved?

File: hicmatrix/test/test_matrixFileHandler.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    # create matrixFileHandler instance with filetype 'cool'
+    pMatrixFile = ROOT + 'Li_et_al_2015.cool'
+    fh = MatrixFileHandler(pFileType='cool', pMatrixFile=pMatrixFile)
+    assert fh is not None
+
+    # load data
+    matrix, cut_intervals, nan_bins, distance_counts, correction_factors = fh.load()
+
+    # TODO(Frederic): figure out how to test matrix.
+
```
Comment (joachimwolff) [inline hicmatrix/test/test_matrixFileHandler.py]: Same question again

**Pull request merged**